### DISCUSSION TITLE
Preserve row selection when updating `ui.aggrid`

### DISCUSSION THREAD

**Discussion by E-H-E-H** [zauberzeug/nicegui#2357] [Ideas / Feature Requests]
### Question

Hi all, 

I use code that receives UDP messages (asynchronous) and updates the ag grid. 
The only thing is, if the user is in the process of selecting rows of data in the grid, an grid data update (as a result of a incoming UDP message) will remove all the marked checkboxes. I think the reason is because the whole grid is being updated instead of the cells afflicted by the data update. 

Is there a way to only update a single cell of an AG grid? of exclude a column in a grid update?
The following simplified code demonstrates the situation:
1. press 'update' button to fill the grid with data.  (simulates an incoming UDP data packet )
2. check some checkboxes. 
3. press 'update' button.  (simulates an incoming UDP data packet )

Eventually, I would like to select certain rows and manipulate them. 

```python
from nicegui import ui
import pandas as pd
import random

grid = ui.aggrid(
    {
        "defaultColDef": {"flex": 1},
        "columnDefs": [
            {"headerName": "Name", "field": "name", 'checkboxSelection': True},
            {"headerName": "Age", "field": "age"},
            {"headerName": "Parent", "field": "parent", "hide": True},
        ],
        "rowData": [],
        "rowSelection": "multiple",
    }
).classes("max-h-40")


class App:
    def __init__(self):
        self.rows = []

    def update(self, df):
        self.rows.clear()
        self.rows.extend(df.to_dict("records"))


app = App()
grid.options["rowData"] = app.rows


def update():
    df = pd.DataFrame(
        [
            {"name": "Alice", "age": random.random(), "parent": "David"},
            {"name": "Bob", "age": random.random(), "parent": "Eve"},
            {"name": "Carol", "age": random.random(), "parent": "Frank"},
        ]
    )
    app.update(df)
    grid.update()


ui.button("Update", on_click=update)
ui.run(reload=False)
```

**Comment by E-H-E-H** [comment]
An possible solution could be to use 'ui.checkbox' (https://nicegui.io/documentation/checkbox) ?
To store the checked rows, then update the grid and restore the checked checkboxes? 
But it sounds like a hassle and not a very stable solution.

**Comment by falkoschindler** [comment]
Hi @E-H-E-H,

This is a tricky problem. It boils down to this:

```py
grid = ui.aggrid({
    'columnDefs': [{'headerName': 'Count', 'field': 'count', 'checkboxSelection': True}],
    'rowData': [{'count': 0}],
})

def count():
    grid.options['rowData'][0]['count'] += 1
    grid.update()

ui.button('Count', on_click=count)
```

Whenever the options are updated, the selection is cleared. I don't know how to avoid it. Has anyone an idea how to improve `ui.aggrid` to (optionally) preserve the selection during updates?

Let's convert this discussion into a feature request.

**Comment by E-H-E-H** [comment]
Hi @falkoschindler ,

Thank you. 

Is it possible to use 'rowNode.setDataValue(col,value)' to update a single value, instead of a whole row? It [mentions](https://www.ag-grid.com/javascript-data-grid/data-update-single-row-cell/#reference-data-setDataValue) only updating the cell value and not the whole grid. 

Furthermore, I found an [example](https://www.ag-grid.com/javascript-data-grid/data-update-transactions/#localised-changes-in-grouped-data) in the AG Grid documentation that does the trick. But I don't know to what level AG Grid is implemented in NiceGUI.

**Reply by falkoschindler** [reply]
Thanks for the hint, @E-H-E-H! It pointed me in the right direction.
I just implemented and merged PR #2413 to support calling row methods in `ui.aggrid`. This way we can update single values without breaking the row selection.

**Reply by E-H-E-H** [reply]
@falkoschindler  Thank you very much! This resolves the issue of updating cell values in existing rows, without losing a checkbox row selection. 

But what are your thoughts about dynamically adding rows without losing the checkbox selection? Like in [this](https://www.ag-grid.com/javascript-data-grid/data-update-transactions/#example-updating-with-transaction-and-groups) example.

**Reply by falkoschindler** [reply]
@E-H-E-H [This](https://www.ag-grid.com/javascript-data-grid/data-update-transactions/#example-updating-with-transaction-and-groups) example looks interesting. We would need to introduce the concept of transactions, which is quite new to me. But maybe we can use it for new methods like `.add_row()`, `.remove_row()` etc.

**Reply by E-H-E-H** [reply]
@falkoschindler,

I was fiddling with some code to solve my earlier question '_But what are your thoughts about dynamically adding rows without losing the checkbox selection?_'

```py
from nicegui import ui
import pandas as pd

df = pd.DataFrame({'name': ['Alice', 'Bob', 'Aritra'],
                   'age': [25, 30, 35],
                   'location': ['Seattle', 'New York', 'Kona']})

grid = ui.aggrid(
    {
        "defaultColDef": {"flex": 1},
        "columnDefs": [
            {"headerName": "Name", "field": "name", 'checkboxSelection': True},
            {"headerName": "Age", "field": "age"},
            {"headerName": "Location", "field": "location"},
        ],
        "rowData": [
            {'name': 'Alice', 'age': 18, 'location': 'New York'},
            {'name': 'Bob', 'age': 21, 'location': 'New York'},
            {'name': 'Carol', 'age': 42, 'location': 'New York'},
        ],
        ':getRowId': '(params) => params.data.name',
        "rowSelection": "multiple",
    }
).classes("w-full max-h-400")

age_val = 99


def add_rows():
    global df
    df1 = pd.DataFrame({
        'name': ['jan', 'kees', 'piet', 'marie'],
        'age': [1, 2, 3, 4],
        'location': ['groningen', 'weesp', 'arnhem', 'nijmegen']
    })
    df = pd.concat([df, df1], ignore_index=True)
    grid.call_api_method('setRowData', df.to_dict("records"))


def update_cell_val():
    global age_val
    age_val = age_val + 1
    grid.run_row_method('Alice', 'setDataValue', 'age', age_val)


ui.button('Update cell value while perserving selection', on_click=update_cell_val)
ui.button("Insert new rows while perserving selection", on_click=add_rows)

grid.update()
ui.run(reload=False)
```

My question is; I think it works, but do you think this is a proper way of doing it? 

(((is there a way to use the full height of the page? I tried 'h-full', but then the whole grid is vanished)))

**Reply by falkoschindler** [reply]
Using `grid.call_api_method('setRowData', ...)` and `grid.call_api_method('setRowData', ...)` looks like a good idea. 👍

To let a NiceGUI element occupy the full screen:
```py
from nicegui import context, ui

# remove the default padding of the page content
context.get_client().content.classes('p-0')

# create a container with the width and height of the screen
with ui.card().classes('w-screen h-screen border-2 border-blue-400') as card:
    ui.label('This container is fullscreen.')
```
